### PR TITLE
Add 'osc import-image' command

### LINKED ISSUE #2769: Add 'osc import-image' command to re-import an image stream's tags
Add a simple CLI command for re-importing an image stream's tags from an external registry.

### PR DESCRIPTION
**Pull request by ncdc**
Fixes #2769

### CODE REVIEW

**Comment by ncdc** [comment]
@smarterclayton @fabianofranz

File: pkg/cmd/cli/cmd/importimage.go [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+		return err
+	}
+
+	imageStreamClient := client.ImageStreams(namespace)
+	stream, err := imageStreamClient.Get(streamName)
+	if err != nil {
+		return err
+	}
+
+	if len(stream.Spec.DockerImageRepository) == 0 {
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/importimage.go]: What if my tags have dockerImageReference or Kind="DockerImage" set

File: pkg/cmd/cli/cmd/importimage.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+		return errors.New("Only image streams with spec.dockerImageRepository set may have images imported")
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/importimage.go]: errors shouldn't be capitalized

File: pkg/cmd/cli/cmd/importimage.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+		updatedStream, err := imageStreamClient.Update(stream)
+		if err != nil {
+			return err
+		}
+
+		resourceVersion = updatedStream.ResourceVersion
+	}
+
+	fmt.Fprintln(out, "Waiting for the import to complete. You may safely cancel at this point - the import will complete in the background.")
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/importimage.go]: Shorten the second sentence - "Waiting for import to complete, CTRL+C to stop waiting"

File: pkg/cmd/cli/cmd/importimage.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+	streamWatch, err := imageStreamClient.Watch(labels.Everything(), fields.SelectorFromSet(fields.Set{"name": stream.Name}), resourceVersion)
+	if err != nil {
+		return err
+	}
+	defer streamWatch.Stop()
+
+	var updatedStream *imageapi.ImageStream
+
+Loop:
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/importimage.go]: Move this into a function - "Loop:" is a code smell for returning early in a function

File: pkg/cmd/cli/cmd/importimage.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+				}
+			case watch.Deleted:
+				return errors.New("the image stream was deleted")
+			default:
+				continue
+			}
+		}
+	}
+
+	if updatedStream != nil {
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/importimage.go]: invert condition, return early

File: pkg/cmd/cli/cmd/importimage.go (hunk 1 above)
Comment (ncdc) [inline pkg/cmd/cli/cmd/importimage.go]: The image importer imports all tags and doesn't currently look at spec.tags. What behavior do you want?

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/importimage.go]: We probably need to support spec tags, ben has a use case around the way we want to manage spec tags (latest points to 2.0, 2.0 points to a repo, but 2.1 may point to a different repo).  Can you get a quick judge of the effort there and work with @bparees?

----- Original Message -----

Comment (ncdc) [inline pkg/cmd/cli/cmd/importimage.go]: Effort probably is minimal - most of the code is still in the import controller but commented out. I'll talk to Ben.

**Comment by ncdc** [comment]
@smarterclayton comments addressed

File: pkg/cmd/cli/cmd/importimage.go [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+have tag and image information imported.`
+
+	importImageExample = `  $ %[1]s import-image mystream`
+)
+
+// NewCmdImportImage implements the OpenShift cli import-image command.
+func NewCmdImportImage(fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "import-image IMAGESTREAM",
+		Short:   "Imports tag and image information from an external Docker image repository.",
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/importimage.go]: Can I not use an image stream tag?

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/importimage.go]: That may be ok, just want to hear an argument for or against

File: pkg/cmd/cli/cmd/importimage.go [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	fmt.Fprintln(out, "Waiting for the import to complete, CTRL+C to stop waiting.")
+
+	updatedStream, err := waitForImport(imageStreamClient, stream.Name, resourceVersion)
+	if err != nil {
+		return fmt.Errorf("unable to determine if the import completed successfully - please run 'osc describe -n %s imagestream/%s' to see if the tags were updated as expected: %v", stream.Namespace, stream.Name, err)
+	}
+
+	fmt.Fprintln(out, "The import completed successfully.\n")
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/importimage.go]: You're mixing two kinds of output - would recommend putting this and the above message on stderr (cmd.Output()) and having describe be the only thing printed to the user

File: pkg/cmd/cli/cmd/importimage.go (hunk 6 above)
Comment (ncdc) [inline pkg/cmd/cli/cmd/importimage.go]: Not sure what you mean. All this does is clear the annotation so the image import controller will process this stream again.

File: pkg/cmd/cli/cmd/importimage.go (hunk 1 above)
Comment (ncdc) [inline pkg/cmd/cli/cmd/importimage.go]: See #2814

File: pkg/cmd/cli/cmd/importimage.go (hunk 7 above)
Comment (ncdc) [inline pkg/cmd/cli/cmd/importimage.go]: Fixed

File: pkg/cmd/cli/cmd/importimage.go (hunk 6 above)
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/importimage.go]: I guess this is fine.

**Comment by smarterclayton** [comment]
LGTM [merge]

**Comment by smarterclayton** [comment]
Add better documentation and description.

**Comment by smarterclayton** [comment]
LGTM [merge]

**Pull request merged**